storage/src/main/java/tech/pegasys/teku/storage/client/RecentChainData.java

```diff
@@ -186,30 +186,43 @@ public StoreTransaction startStoreTransaction() {
    * @param slot the new best slot
    */
   public void updateBestBlock(Bytes32 root, UnsignedLong slot) {
+    final Optional<SignedBlockAndStateAndSlot> originalChainHead = chainHead;
+    store
+        .retrieveBlockAndState(root)
+        .thenApply(
+            headBlockAndState ->
+                headBlockAndState
+                    .map(head -> SignedBlockAndStateAndSlot.create(head, slot))
+                    .orElseThrow(
+                        () ->
+                            new IllegalStateException(
+                                String.format(
+                                    "Unable to update best block as of slot %s.  Block is unavailable: %s.",
+                                    slot, root))))
+        .thenAccept(headBlock -> updateChainHead(originalChainHead, headBlock))
+        .reportExceptions();
+  }
+
+  private void updateChainHead(
+      final Optional<SignedBlockAndStateAndSlot> originalHead,
+      final SignedBlockAndStateAndSlot newChainHead) {
     synchronized (this) {
-      final SignedBeaconBlock newBestBlock = store.getSignedBlock(root);
-      final BeaconState newBestState = store.getBlockState(root);
-      if (newBestBlock == null || newBestState == null) {
-        LOG.warn(
-            "Unable to update best block (slot={}, root={}). Corresponding {} unavailable",
-            slot,
-            root,
-            newBestBlock == null ? "block" : "state");
+      if (!chainHead.equals(originalHead)) {
+        // The chain head has been updated while we were waiting for the newChainHead
+        // Skip this update to avoid accidentally regressing the chain head
+        LOG.debug("Skipping best block update to avoid potential rollback of the best block.");
```

Comment:
I don't think this works.  Say we have:

* Update 1 starts, grabs original head as slot 1, trying to set slot to 2.
* Update 2 starts, grabs original head as slot 1, trying to set slot to 3.
* Update 1 completes and sets head to 2
* Update 2 will now be skipped because it expects head to be 1 but it's changed. We then wind up with an outdated head (but at least we don't go backwards).

I wonder if in `RecentChainData` we should just return the `SafeFuture`.  `ForkChoiceHead` is the key place it gets called from and it probably needs to call `join` so it all happens in the `synchronized` block.  Kind of just pushing the problem down the road unfortunately but I'm struggling to see how else to do it.  Ultimately we need to ensure there is only one thing updating head at a time.

Reply:
Right - the current logic will prevent us from rolling back the chain head, but it doesn't guarantee we get the most up-to-date chain head.

I think as long as we're taking the chain head from `RecentChainData`  it shouldn't matter if we miss an update on rare occasions should it?  We update the head periodically so it should advance on the next call that completes.

Reply:
It makes me nervous and may mean we generate attestations pointing to an outdated head but yeah it probably works -especially since we very likely still have the state in cache since we would have either created it by processing a block or processed an attestation that needed it.

I think it would be well worth printing an `INFO` level message each time we skip an update though. If it happens a lot we should be worried and I think it's useful to know how often it is happening.